autoload/ale/completion.vim

```diff
@@ -183,52 +185,71 @@ function! ale#completion#RestoreCompletionOptions() abort
+
 function! ale#completion#OmniFunc(findstart, base) abort
     if a:findstart
-        let l:line = b:ale_completion_info.line
-        let l:column = b:ale_completion_info.column
-        let l:regex = s:GetFiletypeValue(s:omni_start_map, &filetype)
-        let l:up_to_column = getline(l:line)[: l:column - 2]
-        let l:match = matchstr(l:up_to_column, l:regex)
-
-        return l:column - len(l:match) - 1
+        return ale#completion#FindCompletionStart()
     else
-        " Parse a new response if there is one.
-        if exists('b:ale_completion_response')
-        \&& exists('b:ale_completion_parser')
-            let l:response = b:ale_completion_response
-            let l:parser = b:ale_completion_parser
+        return ale#completion#PollCompletionResults()
+    endif
+endfunction
 
-            unlet b:ale_completion_response
-            unlet b:ale_completion_parser
+function! ale#completion#QueryDone() abort
+    let s:last_done_pos = getcurpos()[1:2]
 
-            let b:ale_completion_result = function(l:parser)(l:response)
-        endif
+    let b:ale_completion_result = ALEProcessCompletionResults(get(b:, 'ale_completion_result', []))
 
-        call s:ReplaceCompletionOptions()
+    call ALEAfterCompletionResults()
+endfunction
 
-        return get(b:, 'ale_completion_result', [])
-    endif
+function! ALEProcessCompletionResults(completion_results) abort
+    " Users may override this function to act on the results however they
+    " see fit. Needs to be not namespaced.
+    return a:completion_results
 endfunction
 
-function! ale#completion#Show(response, completion_parser) abort
-    if ale#util#Mode() isnot# 'i'
+function! ALEAfterCompletionResults() abort
+    " Users may override this function to act on the results however they
+    " see fit. Needs to be not namespaced.
+    if !ale#util#Mode() is# 'i'
         return
     endif
 
-    " Set the list in the buffer, temporarily replace omnifunc with our
-    " function, and then start omni-completion.
-    let b:ale_completion_response = a:response
-    let b:ale_completion_parser = a:completion_parser
-    " Replace completion options shortly before opening the menu.
-    call s:ReplaceCompletionOptions()
-
-    call timer_start(0, {-> ale#util#FeedKeys("\<Plug>(ale_show_completion_menu)")})
+    " If the completion events are not managed by ALE, don't mess with
+    " the omnifunc option. Otherwise, replace completion options shortly
+    " before opening the menu.
+    if g:ale_completion_enabled
+        call s:ReplaceCompletionOptions()
+        call timer_start(0, {-> ale#util#FeedKeys("\<Plug>(ale_show_completion_menu)")})
+    endif
 endfunction
 
 function! s:CompletionStillValid(request_id) abort
     let [l:line, l:column] = getcurpos()[1:2]
 
+    return has_key(b:, 'ale_completion_info')
+    \&& b:ale_completion_info.request_id == a:request_id
+    \&& b:ale_completion_info.line == l:line
+    \&& b:ale_completion_info.column == l:column
```

Comment:
You need to remove the `return` statement below. Something will have to be done about checking that the mode is `'i'` still, for insert mode. Make sure we still check that the mode is `'i'` (Using the Mode wrapper function so we can test it.) before using `feedkeys()` with `<C-x><C-o>`.

Reply:
It is a mistake that there are two return statements,

but the name `CompletionStillValid` speaks to me that it checks if completion is still valid, not if Vim is still in insert mode or if ALE is supposed to show this in the screen. The outside source shouldn't be bound by this behavior.

Reply:
Yes. Remove the `return` statement below this one, as I said.